### PR TITLE
MOSIP-13091 : Added Quality Classification Tags

### PR DESCRIPTION
**Pull request by yaswanths2**

### CODE REVIEW

File: registration-processor/pre-processor/registration-processor-quality-checker-stage/src/main/java/io/mosip/registration/processor/quality/checker/stage/QualityCheckerStage.java [hunk 1]
```diff
@@ -360,11 +415,11 @@ public MessageDTO process(MessageDTO object) {
...
+					regId, RegistrationStatusCode.FAILED.toString() + e.getMessage()
+							+ org.apache.commons.lang3.exception.ExceptionUtils.getStackTrace(e));
 			registrationStatusDto.setStatusCode(RegistrationStatusCode.PROCESSING.name());
-			registrationStatusDto.setStatusComment(
-					trimExceptionMsg.trimExceptionMessage(StatusUtil.PACKET_MANAGER_EXCEPTION.getMessage() + e.getMessage()));
+			registrationStatusDto.setStatusComment(trimExceptionMsg
+					.trimExceptionMessage(StatusUtil.PACKET_MANAGER_EXCEPTION.getMessage() + e.getMessage()));
 			registrationStatusDto.setSubStatusCode(StatusUtil.PACKET_MANAGER_EXCEPTION.getCode());
 			registrationStatusDto.setLatestTransactionStatusCode(
 					registrationStatusMapperUtil.getStatusCode(RegistrationExceptionTypeCode.PACKET_MANAGER_EXCEPTION));
```
Comment (vishwa-vyom) [inline registration-processor/pre-processor/registration-processor-quality-checker-stage/src/main/java/io/mosip/registration/processor/quality/checker/stage/QualityCheckerStage.java:353]: Please include the unit test cases for the new classification logic

File: registration-processor/pre-processor/registration-processor-quality-checker-stage/src/main/java/io/mosip/registration/processor/quality/checker/stage/QualityCheckerStage.java [hunk 2]
```diff
@@ -146,12 +152,36 @@
...
 	private RegistrationExceptionMapperUtil registrationStatusMapperUtil;
 
+	/**
+	 * Below quality classifications map should contain proper quality
+	 * classification name and quality range, any overlap of the quality range will
+	 * result in a random behaviour of tagging. In range, upper and lower values are
+	 * inclusive.
+	 */
+	
+	@Value("#{${mosip.regproc.packet.classifier.tagging.quality.ranges:{'Poor':'0-29','Good':'30-69','Very Good':'70-100'}}}")
```
[the remark this answers is not in the document]
Comment (vishwa-vyom) [reply registration-processor/pre-processor/registration-processor-quality-checker-stage/src/main/java/io/mosip/registration/processor/quality/checker/stage/QualityCheckerStage.java]: Also change the ranges as Poor, Average, Good

**Review by vishwa-vyom — CHANGES_REQUESTED** [review]

File: registration-processor/pre-processor/registration-processor-quality-checker-stage/src/main/java/io/mosip/registration/processor/quality/checker/stage/QualityCheckerStage.java [hunk 3]
```diff
@@ -62,17 +69,13 @@
...
-		"io.mosip.registration.processor.rest.client.config", 
-		"io.mosip.registration.processor.packet.storage.config",
-		"io.mosip.registration.processor.packet.manager.config", 
-		"io.mosip.kernel.idobjectvalidator.config",
+		"io.mosip.registration.processor.quality.checker.config", "io.mosip.registration.processor.stages.config",
+		"io.mosip.registrationprocessor.stages.config", "io.mosip.registration.processor.status.config",
+		"io.mosip.registration.processor.rest.client.config", "io.mosip.registration.processor.packet.storage.config",
+		"io.mosip.registration.processor.packet.manager.config", "io.mosip.kernel.idobjectvalidator.config",
 		"io.mosip.registration.processor.core.kernel.beans", "io.mosip.kernel.biosdk.provider.impl" })
 public class QualityCheckerStage extends MosipVerticleAPIManager {
```
[the remark this answers is not in the document]
Comment (yaswanths2) [reply registration-processor/pre-processor/registration-processor-quality-checker-stage/src/main/java/io/mosip/registration/processor/quality/checker/stage/QualityCheckerStage.java]: changed project, package,class, artifact id, and module names

File: registration-processor/pre-processor/registration-processor-quality-checker-stage/src/main/java/io/mosip/registration/processor/quality/checker/stage/QualityCheckerStage.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (yaswanths2) [reply registration-processor/pre-processor/registration-processor-quality-checker-stage/src/main/java/io/mosip/registration/processor/quality/checker/stage/QualityCheckerStage.java]: changed classifer to Poor, Average and Good. changed propert name

File: registration-processor/pre-processor/registration-processor-quality-checker-stage/src/main/java/io/mosip/registration/processor/quality/checker/stage/QualityCheckerStage.java [hunk 4]
```diff
@@ -146,12 +152,36 @@
...
 
+	/**
+	 * Below quality classifications map should contain proper quality
+	 * classification name and quality range, any overlap of the quality range will
+	 * result in a random behaviour of tagging. In range, upper and lower values are
+	 * inclusive.
+	 */
+	
+	@Value("#{${mosip.regproc.packet.classifier.tagging.quality.ranges:{'Poor':'0-29','Good':'30-69','Very Good':'70-100'}}}")
+	private Map<String, String> qualityClassficaticationsRangeMap;
```
[the remark this answers is not in the document]
Comment (yaswanths2) [reply registration-processor/pre-processor/registration-processor-quality-checker-stage/src/main/java/io/mosip/registration/processor/quality/checker/stage/QualityCheckerStage.java]: corrected spelling

File: registration-processor/pre-processor/registration-processor-quality-checker-stage/src/main/java/io/mosip/registration/processor/quality/checker/stage/QualityCheckerStage.java [hunk 5]
```diff
@@ -165,6 +195,18 @@
 	@Autowired
 	private BioAPIFactory bioApiFactory;
 
+	@PostConstruct
+	private void generateParsedAgeGroupRangeMap() {
```
[the remark this answers is not in the document]
Comment (yaswanths2) [reply registration-processor/pre-processor/registration-processor-quality-checker-stage/src/main/java/io/mosip/registration/processor/quality/checker/stage/QualityCheckerStage.java]: modified method name

File: registration-processor/pre-processor/registration-processor-quality-checker-stage/src/main/java/io/mosip/registration/processor/quality/checker/stage/QualityCheckerStage.java [hunk 6]
```diff
@@ -146,12 +152,36 @@
...
+	@Value("${mosip.regproc.packet.classifier.tagging.quality.prefix:Quality_}")
+	private String qualityTagPrefix;
+
+	private static String RANGE_DELIMITER = "-";
+
+	/**
+	 * Filter qualityClassficaticationsRangeMap using delimiter and store into
+	 * parsedQualityRangemap using @PostConstruct
+	 */
+	private Map<String, int[]> parsedQualityRangemap;
```
[the remark this answers is not in the document]
Comment (yaswanths2) [reply registration-processor/pre-processor/registration-processor-quality-checker-stage/src/main/java/io/mosip/registration/processor/quality/checker/stage/QualityCheckerStage.java]: modified var name

File: registration-processor/pre-processor/registration-processor-quality-checker-stage/src/main/java/io/mosip/registration/processor/quality/checker/stage/QualityCheckerStage.java [hunk 7]
```diff
@@ -146,12 +152,36 @@
...
+	 * Below quality classifications map should contain proper quality
+	 * classification name and quality range, any overlap of the quality range will
+	 * result in a random behaviour of tagging. In range, upper and lower values are
+	 * inclusive.
+	 */
+	
+	@Value("#{${mosip.regproc.packet.classifier.tagging.quality.ranges:{'Poor':'0-29','Good':'30-69','Very Good':'70-100'}}}")
+	private Map<String, String> qualityClassficaticationsRangeMap;
+
+	@Value("${mosip.regproc.packet.classifier.tagging.quality.prefix:Quality_}")
```
[the remark this answers is not in the document]
Comment (yaswanths2) [reply registration-processor/pre-processor/registration-processor-quality-checker-stage/src/main/java/io/mosip/registration/processor/quality/checker/stage/QualityCheckerStage.java]: changed the default value

File: registration-processor/pre-processor/registration-processor-quality-checker-stage/src/main/java/io/mosip/registration/processor/quality/checker/stage/QualityCheckerStage.java [hunk 8]
```diff
@@ -237,59 +282,70 @@ public MessageDTO process(MessageDTO object) {
...
-						description.setMessage(PlatformErrorMessages.BIOMETRIC_QUALITY_CHECK_FAILED.getMessage());
-						break;
-					} else {
-						scoreCounter++;
-					}
+					birArray[0] = bir;
+					float[] qualityScoreresponse = getBioSdkInstance(biometricType).getSegmentQuality(birArray, null);
+
+					float score = Float.valueOf(qualityScoreresponse[0]);
+					String tagName = qualityTagPrefix.concat(bir.getBdbInfo().getType().get(0).value());
```
[the remark this answers is not in the document]
Comment (yaswanths2) [reply registration-processor/pre-processor/registration-processor-quality-checker-stage/src/main/java/io/mosip/registration/processor/quality/checker/stage/QualityCheckerStage.java]: changes made as suggested

File: registration-processor/pre-processor/registration-processor-quality-checker-stage/src/main/java/io/mosip/registration/processor/quality/checker/stage/QualityCheckerStage.java (hunk 1 above)
Comment (yaswanths2) [reply registration-processor/pre-processor/registration-processor-quality-checker-stage/src/main/java/io/mosip/registration/processor/quality/checker/stage/QualityCheckerStage.java:353]: Added UT and tested

File: registration-processor/pre-processor/registration-processor-quality-classifier-stage/src/test/java/io/mosip/registration/processor/quality/qualifier/stage/test/QualityClassifierStageTest.java [hunk 9]
```diff
@@ -381,54 +415,54 @@ public void testQualityCheckfailureException() throws BiometricException {
...
 		dto.setRid("1234567890");
-		MessageDTO result = qualityCheckerStage.process(dto);
+		MessageDTO result = qualityClassifierStage.process(dto);
 
 		assertFalse(result.getIsValid());
 	}
 
 
 	@Test
 	public void testJsonProcessingException() throws ApisResourceAccessException, IOException, PacketManagerException, JsonProcessingException {
```
Comment (vishwa-vyom) [inline registration-processor/pre-processor/registration-processor-quality-classifier-stage/src/test/java/io/mosip/registration/processor/quality/qualifier/stage/test/QualityClassifierStageTest.java]: You have to add unit test cases to check if the tagging method is called with proper tag created based on the score. Have 3 test cases one with 
1. All bio types are returning same (Good for all 3)
2. returns 2 combinations, (1 Good and 2 Poor), 
3. All 3 bio types are giving 3 different classifications (1 Good, 1 Poor, 1 Average), lets check if the tags are created properly based on the input score

https://github.com/mosip/registration/blob/develop/registration-processor/pre-processor/registration-processor-packet-classifier-stage/src/test/java/io/mosip/registration/processor/stages/packetclassifier/tagging/impl/AgeGroupTagGeneratorTest.java

Refer below for ArgumentCaptor to get the tags that is passed by QualityCheckerStage to addOrUpdate tags method
https://github.com/mosip/registration/blob/develop/registration-processor/registration-processor-core/src/test/java/io/mosip/registration/processor/core/eventbus/KafkaMosipEventBusTest.java

**Review by vishwa-vyom — CHANGES_REQUESTED** [review]

File: registration-processor/pre-processor/registration-processor-quality-classifier-stage/src/test/java/io/mosip/registration/processor/quality/qualifier/stage/test/QualityClassifierStageTest.java (hunk 9 above)
Comment (yaswanths2) [reply registration-processor/pre-processor/registration-processor-quality-classifier-stage/src/test/java/io/mosip/registration/processor/quality/qualifier/stage/test/QualityClassifierStageTest.java]: Added UT for POOR, AVG,GOOD tags, but couldnt write it for mix and match

File: registration-processor/pre-processor/registration-processor-quality-classifier-stage/src/main/java/io/mosip/registration/processor/quality/classifier/stage/QualityClassifierStage.java [hunk 10]
```diff
@@ -463,32 +391,50 @@ public MessageDTO process(MessageDTO object) {
...
-		}
-		return 0;
-	}
 
 	private iBioProviderApi getBioSdkInstance(BiometricType biometricType) throws BiometricException {
 		iBioProviderApi bioProvider = bioApiFactory.getBioProvider(biometricType, BiometricFunction.QUALITY_CHECK);
 		return bioProvider;
 	}
+	
+	public Map<String, String> getQualityTags(List<BIR> birs) throws BiometricException{
```
Comment (vishwa-vyom) [inline registration-processor/pre-processor/registration-processor-quality-classifier-stage/src/main/java/io/mosip/registration/processor/quality/classifier/stage/QualityClassifierStage.java]: Why is this method public ?

**Review by vishwa-vyom — CHANGES_REQUESTED** [review]
For the sake of unit test cases, we cannot declare a method as public. You have to use Argument Capture to test these cases.

**Review by vishwa-vyom — APPROVED** [review]

**Pull request merged**